File: lol/src/test/kotlin/com/your/lol/acceptance/SummonerStatsAcceptanceTest.java
package com.your.lol.acceptance;

import com.google.gson.reflect.TypeToken;
import com.your.lol.dto.riot.MatchDto;
import com.your.lol.support.JsonFileConverter;
import io.restassured.response.ValidatableResponse;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;

import java.util.List;

import static org.hamcrest.Matchers.equalTo;
import static org.mockito.BDDMockito.given;

public class SummonerStatsAcceptanceTest extends AcceptanceTest {

    @Test
    void findStatisticsBySummonerName() {
        // given
        List<MatchDto> matchDtos = createTestData();
        given(webClientFacade.requestRiotStatistics("테스트 소환사"))
                .willReturn(matchDtos);

        // when
        ValidatableResponse response = get("/api/summoner/" + "테스트 소환사");

        // then
        response.statusCode(HttpStatus.OK.value())
                .body("winRate", equalTo(0.5));
    }

    private List<MatchDto> createTestData() {
        return JsonFileConverter.fromJsonFile(
                "match-dtos.json",
                new TypeToken<>() {
                });
    }
}

File: lol/src/test/kotlin/com/your/lol/acceptance/SummonerStatsAcceptanceTest.kt
package com.your.lol.acceptance

import com.your.lol.dto.statistics.ChampionDataDto
import com.your.lol.dto.statistics.StatsDto
import io.mockk.every
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito
import org.mockito.BDDMockito.given

class SummonerStatsAcceptanceTest : AcceptanceTest() {

    @Test
    fun `소환사 이름을 검색하여 전적 통계를을 산출한다`() {
        // given
        val matchDtos = createTestData("match-dtos.json")
        given(webClientFacade.requestRiotStatistics("테스트 소환사 GG"))
                .willReturn(matchDtos)

        // when
        val response = get("/api/summoner/${"테스트 소환사 GG"}")
        val statsDto = response!!.extract().`as`(StatsDto::class.java)
        val jinx: ChampionDataDto = findChampionByName(statsDto, "Jinx")
        val riven: ChampionDataDto = findChampionByName(statsDto, "Riven")

        // then
        Assertions.assertAll(
                assertCloseTo<Double>(statsDto.winRate, 0.666, 0.001),
                assertEquals(jinx.matchCount, 1),
                assertEquals(jinx.winRate, 1.0),
                assertCloseTo<Double>(jinx.averageKda, 3.6, 0.01),
                assertEquals(riven.matchCount, 2),
                assertEquals(riven.winRate, 0.5),
                assertCloseTo<Double>(riven.averageKda, 5.97, 0.01)
        )
    }
}
